homeassistant/helpers/config_entry_oauth2_flow.py

```diff
@@ -199,12 +201,15 @@ async def _token_request(self, data: dict) -> dict:
 
         _LOGGER.debug("Sending token request to %s", self.token_url)
         resp = await session.post(self.token_url, data=data)
-        if resp.status >= 400 and _LOGGER.isEnabledFor(logging.DEBUG):
-            body = await resp.text()
-            _LOGGER.debug(
-                "Token request failed with status=%s, body=%s",
-                resp.status,
-                body,
+        if resp.status >= 400:
+            try:
+                error_response = await resp.json()
+            except (ClientError, JSONDecodeError):
+                error_response = {}
+            error_code = error_response.get("error", "unknown")
+            error_description = error_response.get("error_description", "unknown error")
+            _LOGGER.error(
+                "Token request failed (%s): %s", error_code, error_description
             )
         resp.raise_for_status()
```

Comment:
We could wrap the `ClientResponseError` in our own exception and add the error code and error description as additional attributes. We already check if the response status is 400 or higher above.

Reply:
Ok I was considering this, these options: given existing use of this code (used in both the config flow or generally for the auth impls token refresh), I was hesitant to introduce a new exception. Re throwing an siohttp exception could work but seems rare in this code base (and one param to the constructor may be hard to get or would be faked).  I considered splitting the two token request use cases here into two calls as well or slightly changing the API to return a response and handle exceptions in the caller. Thoughts?

Reply:
The case I am worried about is where this function is called for normal token refresh and integrations catch aiohttp exceptions directly, like on async setup, outside of config flow.

We could also upgrade all those code paths with more explicit oauth exceptions if we want to fully solve this.

Reply:
The token refresh is handled by the Home Assistant oauth session so we can decide what exceptions should be handled by callers for that api.

Reply:
I suggest we go ahead with this PR and then improve and add the new exception separately.